atomicapp/nulecule/main.py

```diff
@@ -0,0 +1,234 @@
+        self.render()
+
+    def run(self, provider_key=None, dry=False):
+        provider_key, provider = self.get_provider(provider_key, dry)
+        for component in self.components:
+            component.run(provider_key, dry)
+
+    def stop(self, provider_key=None):
+        provider_key, provider = self.get_provider(provider_key)
+        # stop the Nulecule application
+        for component in self.components:
+            component.stop(provider_key)
+
+    def uninstall(self):
+        # uninstall the Nulecule application
+        for component in self.components:
+            component.uninstall()
+
+    def load_config(self, config={}):
+        super(Nulecule, self).load_config(config=config)
+        for component in self.components:
+            _config = {}
+            _config[GLOBAL_CONF] = copy.deepcopy(
+                self.config.get(GLOBAL_CONF)) or {}
+            if self.namespace != GLOBAL_CONF:
+                _config[GLOBAL_CONF].update(
+                    self.config.get(self.namespace) or {})
+            _config[component.name] = copy.deepcopy(
+                self.config.get(component.name)) or {}
+            component.load_config(config=_config)
+            self.merge_config(self.config, component.config)
+
+    def load_components(self):
+        components = []
+        for node in self.graph:
+            node_name = node['name']
+            component = NuleculeComponent(
+                node_name, self.basepath, node.get('source'),
+                node.get('params'), node.get('artifacts'))
+            component.load()
+            components.append(component)
+        self.components = components
+
+    def render(self):
+        for component in self.components:
+            component.render()
+
+
+class NuleculeComponent(NuleculeBase):
+    """
+    Represents a component in a Nulecule application. It receives props
+    from it's parent and can add new props and override props at it's local
+    scope. It does not have direct access to props of sibling Nulecule
+    components, but can request the value of sibling's property from it's
+    parent.
+    """
+    def __init__(self, name, basepath, source=None, params=None,
+                 artifacts=None, config=None):
+        self.name = self.namespace = name
+        self.basepath = basepath
+        self.source = source
+        self.params = params or []
+        self.artifacts = artifacts
+        self._app = None
+
+    def load(self):
+        dest = os.path.join(EXTERNAL_APP_DIR, self.name)
+        if not self.artifacts:
+            self.load_external_application(dest)
+
+    def run(self, provider_key, dry=False):
+        if self._app:
+            self._app.run(provider_key, dry)
+            return
+        provider_key, provider = self.get_provider(provider_key, dry)
+        provider.artifacts = self.get_artifact_paths_for_provider(
+            provider_key)
+        provider.init()
+        if not dry:
```

Comment:
Always do `deploy`...provider methods should take care of dry run itself..dry run should print out commands that would be run if it was real run..